privacy-protections/fingerprinting/helpers/tests.js

```diff
@@ -0,0 +1,1292 @@
+            vertexPosBuffer.itemSize = 3
+            vertexPosBuffer.numItems = 3
+            const program = gl.createProgram()
+            const vshader = gl.createShader(gl.VERTEX_SHADER)
+            gl.shaderSource(vshader, vShaderTemplate)
+            gl.compileShader(vshader)
+            const fshader = gl.createShader(gl.FRAGMENT_SHADER)
+            gl.shaderSource(fshader, fShaderTemplate)
+            gl.compileShader(fshader)
+            gl.attachShader(program, vshader)
+            gl.attachShader(program, fshader)
+            gl.linkProgram(program)
+            gl.useProgram(program)
+            program.vertexPosAttrib = gl.getAttribLocation(program, 'attrVertex')
+            program.offsetUniform = gl.getUniformLocation(program, 'uniformOffset')
+            gl.enableVertexAttribArray(program.vertexPosArray)
+            gl.vertexAttribPointer(program.vertexPosAttrib, vertexPosBuffer.itemSize, gl.FLOAT, !1, 0, 0)
+            gl.uniform2f(program.offsetUniform, 1, 1)
+            gl.drawArrays(gl.TRIANGLE_STRIP, 0, vertexPosBuffer.numItems)
+
+            return gl.canvas.toDataURL();
+        }
+    },
+
+    // optional - all props
+    {
+        id: 'window',
+        category: 'all-props',
+        getValue: () => {
+            const testedProps = tests.filter(t => t.category === 'window').map(t => t.id.match('window\.([^.(]*).*')[1]);
+            const ignoredTypes = ['object', 'function', 'undefined'];
+            const result = {};
+
+            Object.keys(window).forEach(propName => {
+                // ignore props that we alredy test
+                if (testedProps.includes(propName)) {
+                    return;
+                }
+
+                const propType = typeof window[propName];
+
+                // ignore props of complex types
+                if (ignoredTypes.includes(propType)) {
+                    return;
+                }
+
+                result[propName] = window[propName];
+            });
+
+            return result;
+        }
+    },
+    {
+        id: 'navigator',
+        category: 'all-props',
+        getValue: () => {
+            const testedProps = tests.filter(t => t.category === 'navigator').map(t => t.id.match('navigator\.([^.(]*).*')[1]);
+            const ignoredTypes = ['object', 'function', 'undefined'];
+            const result = {};
+
+            Object.keys(navigator).forEach(propName => {
+                // ignore props that we alredy test
+                if (testedProps.includes(propName)) {
+                    return;
+                }
+
+                const propType = typeof navigator[propName];
+
+                // ignore props of complex types
+                if (ignoredTypes.includes(propType)) {
+                    return;
+                }
+
+                result[propName] = navigator[propName];
+            });
+
+            return result;
+        }
+    },
+    {
+        id: 'screen',
```

Comment:
Is this something we should consolidate into a stand alone function? Seems we could pass an object name and run through this, so we could more easily add additional objects to enumerate in the future if we wanted to beyond just screen and navigator.

Reply:
Ah, really good tip, that's clearly almost identical code. I extracted it to a function AND while playing with it found out how to list even more properties (my use of `Object.keys` was wrong). I also decided to include properties with complex values but just replace the values with simple placeholders ('~object~').

<img width="1728" alt="moar" src="https://user-images.githubusercontent.com/985504/101064937-d7fc4c00-3594-11eb-96c5-c76099b32619.png">